feat(ui): Added releasesV2 detail page (wip)

matejminar opened the pull request:
This PR adds the releasesV2 detail page.

![Screenshot 2020-03-10 at 10 23 30](https://user-images.githubusercontent.com/9060071/76330886-ab115600-62ee-11ea-87e3-1996e3351b94.png)

Work in progress behind feature flag.

This is only 'Overview' tab and many components on it are still waiting for API to be finished.

The rest of the tabs will be in separate PRs.
I would like to merge this batch even though it's not completely finished yet because otherwise, this will get huge and hard to review.

matejminar commented on src/sentry/static/sentry/app/components/stream/group.jsx:
```diff
@@ -1,4 +1,5 @@
 import $ from 'jquery';
+// eslint-disable-next-line no-restricted-imports
```
It would have been better to move this file to TS and remove `reflexbox` and `jquery` on the way, but I didn't want to include it in the scope of this PR

matejminar commented on src/sentry/static/sentry/app/components/stream/group.jsx:
```diff
@@ -1,4 +1,5 @@
 import $ from 'jquery';
+// eslint-disable-next-line no-restricted-imports
 import {Flex, Box} from 'reflexbox';
```
It would have been better to move this file to TS and remove `reflexbox` and `jquery` on the way, but I didn't want to include it in the scope of this PR

priscilawebdev commented on src/sentry/static/sentry/app/views/releasesV2/detail/overview/commitAuthorBreakdown.tsx:
```diff
@@ -0,0 +1,143 @@
+    // group commits by author
+    const groupedAuthorCommits = this.state.commits?.reduce(
+      (authorCommitsAccumulator, commit) => {
+        const email = commit.author?.email ?? 'undefined';
```
Are you sure about that? is it possible for the authorCommitsAccumulator to have `undefined`?

matejminar replied:
It's a key to group all commits that don't have known authors together. I change the string key into 'unknown' to avoid any confusion.

priscilawebdev commented on src/sentry/static/sentry/app/views/releasesV2/detail/overview/commitAuthorBreakdown.tsx:
```diff
@@ -0,0 +1,143 @@
+      <Wrapper>
+        <SectionHeading>{t('Commit Author Breakdown')}</SectionHeading>
+        {sortedAuthorsByNumberOfCommits.map(({commitCount, author}) => (
+          <AuthorLine key={author?.email}>
```
and if the author is undefined or null, the probability of having a duplicate undefined key is high. I always avoid to have the index as a key, but maybe in this case you should use it

matejminar replied:
they are grouped by email before - so the emails will be unique and there will be only one key named 'unknown'

priscilawebdev commented on src/sentry/static/sentry/app/views/releasesV2/detail/overview/healthChart.tsx:
```diff
@@ -0,0 +1,113 @@
+import withOrganization from 'app/utils/withOrganization';
+import {Client} from 'app/api';
+
+// VERY ROUGH MOCK, MORE LIKE VISUAL PLACEHOLDER, this will be changed completely
```
👍 so I will review it later on

priscilawebdev commented on src/sentry/static/sentry/app/views/releasesV2/detail/overview/issues.tsx:
```diff
@@ -0,0 +1,93 @@
+      start: '',
+      end: '',
+      environment: [''],
+    } as const;
```
do we really need to use casting here? I believe that it is always better to avoid them

priscilawebdev commented on src/sentry/static/sentry/app/views/releasesV2/detail/overview/totalCrashFreeUsers.tsx:
```diff
@@ -0,0 +1,79 @@
+          <InnerRow>
+            <Text bold>March 7</Text>
+            <Text bold right>
+              4.8k users
```
I believe this is a mock :)

matejminar replied:
It's a wip mock

matejminar commented on src/sentry/static/sentry/app/utils/formatters.tsx:
```diff
@@ -4,7 +4,7 @@ import {Release} from '@sentry/release-parser';
 import {CommitAuthor, User} from 'app/types';
 
-export function userDisplayName(user: User | CommitAuthor): string {
+export function userDisplayName(user: User | CommitAuthor, includeEmail = true): string {
```
I believe the correct type is being inferred when assigning primitives like boolean. This seems to me just like a pure duplication.

matejminar commented on src/sentry/static/sentry/app/views/releasesV2/detail/index.tsx:
```diff
@@ -35,60 +50,74 @@ class ReleasesV2Detail extends AsyncView<Props, State> {
 
-    if (!dummy.length) {
-      return this.renderEmpty();
+  renderError(error: Error, disableLog = false, disableReport = false) {
```
@getsentry/app-frontend any opinions here? I think both options are valid, but we should stay consistent.

markstory replied:
I think being explicit makes signatures more consistent. But it is also a bit redundant as typescript can infer the type easily.

priscilawebdev replied:
this is something that we could bring to the TSC meeting today

markstory commented on src/sentry/static/sentry/app/components/version.tsx:
```diff
@@ -50,6 +50,10 @@ type Props = {
+  /**
+   * Use releases v2 (for example when linking to detail page)
+   */
+  v2?: boolean;
```
If an organization has releases-v2 should all links go to the v2 pages? If so you could wrap the Version component with `withOrganization()` so you have access to the current organization and not need to update all the callsites.

matejminar replied:
I will definitely do it this way - it will be a part of a separate PR. This is a temporary solution (demo purposes) - for now, you go to releasesV2 detail page only from releasesV2 listing.

markstory commented on src/sentry/static/sentry/app/views/releasesV2/detail/breadcrumbs.tsx:
```diff
@@ -0,0 +1,56 @@
+const StyledIcon = styled(IconChevron)`
+  color: ${p => p.theme.gray2};
+  height: 12px;
+  width: 12px;
```
Could use the standard icon sizes?

markstory commented on src/sentry/static/sentry/app/views/releasesV2/detail/overview/commitAuthorBreakdown.tsx:
```diff
@@ -0,0 +1,143 @@
+
+        return authorCommitsAccumulator;
+      },
+      {} as {[key: string]: {author: User | undefined; commitCount: number}}
```
Should this be a named type?

matejminar replied:
Sure I refactored it - I was thinking about it too, but it was only used there and didn't wrap to multiple lines so... well you know the rest

markstory commented on src/sentry/static/sentry/app/views/releasesV2/detail/overview/issues.tsx:
```diff
@@ -0,0 +1,93 @@
+      version: 2,
+      name: `${t('Release')} ${formatVersion(version)}`,
+      fields: ['title', 'count(id)', 'event.type', 'user', 'last_seen'],
+      query: `release:${version}`,
```
Do you only want errors?

matejminar replied:
This query is more like a temporary placeholder just so it links to the discover. I will ask @fdrabek to decide what the final discover result should look like and then I will probably bug you with it a little bit :)

matejminar commented:
Thanks for unblocking me. I have addressed all the feedback and I've merged it so that the product can play with it early and shoot a video for show and tell.

Any additional feedback is welcome and will be addressed in separate PRs.
